lib/spack/spack/extensions.py

```diff
@@ -11,10 +11,65 @@
 import types
 
 import llnl.util.lang
-import llnl.util.tty as tty
+import spack.cmd
 import spack.config
+from spack.error import SpackError
 
-extension_regexp = re.compile(r'spack-([\w]*)')
+_command_paths = []
+_extension_regexp = re.compile(r'spack-([\w]*)')
+_extension_command_map = None
+
+
+class CommandNotFoundError(SpackError):
+    """Exception class thrown when a requested command is not recognized as
+    such.
+    """
+    def __init__(self, cmd_name):
+        super(CommandNotFoundError, self).__init__(
+            '{0} is not a recognized Spack command or extension command.'
+            .format(cmd_name),
+            'Known commands: {0}'.format(' '.join(spack.cmd.all_commands())))
+
+
+class ExtensionNamingError(SpackError):
+    """Exception class thrown when a configured extension does not follow
+    the expected naming convention.
+    """
+    def __init__(self, path):
+        super(ExtensionNamingError, self).__init__(
+            '{0} does not match the format for a Spack extension path.'
+            .format(path))
+
+
+def _init_extension_command_map():
+    """Return the list of paths where to search for command files."""
+    global _extension_command_map
+    if _extension_command_map is None:
+        _extension_command_map = {}
+        extension_paths = spack.config.get('config:extensions') or []
+        for path in extension_paths:
+            extension = extension_name(path)
+            command_path = os.path.join(path, extension, 'cmd')
+            _command_paths.append(command_path)
+            commands = spack.cmd.find_commands(command_path)
+            _extension_command_map.update(
+                dict((command, path) for command in
```

Comment:
(Minor) You don't have to update with a `dict`, you can do it with a generator of tuples (i.e. just remove the `dict` constructor invocation and keep everything else).

Reply:
Is there a reason this doesn't just map `command` to `command_path`? That appears to be recomputed in `load_command_extension`.

Also: I think if that's the case, then `_command_paths` can be directly computed from the map (IMO there should be a `command_paths()` method which invokes `_init_extension_command_map` and then returns `_extension_command_map.values()`.

Reply:
@scheibel Sorry, just getting back to this after deadlines and vacation.
* Removed `dict` as suggested.
* I'm really not sure what you mean here. An extension and a command are two different things: an extension effectively forms the namespace, and the command is not necessarily the same. For instance, I might provide two commands, `spack dev` and `spack findext` as part of the `spack-dev` extension package. Here, `dev` is both the name of the extension and the name of _one_ of the two commands provided by the extension. `_init_extension_command_map()` only does its work once (except as part of the tests). `load_command_extension()` locates the correct `command_path` corresponding to  the desired `command` in `_extension_command_map` and uses it to compute the extension name needed in turn for the namespaced module name. The mapped value for each command is the top level directory of the extension, not the full `command_path`. `command_path` is calculated in `_init_extension_command_map()` in order to be able to locate the commands for each extension, but is not saved in `_extension_command_map` (see why in a moment). The `command_path` for the one command we actually decide to load is recalculated in order to actually load the command module. We don't save it when the map is constructed because we _also_ need to know the name of the extension explicitly for the module namespace, which can be calculated from the top level extension path via regex, but not necessarily via the full command path. Given that we only actually need to calculate the extension name and recalculate the command path once, I decided it wasn't worth saving multiple items per path, when the extra items would only be needed for one command under most circumstances.
  TL;DR: one of possibly many`command_path` values is recomputed once which _could_ have been saved, but wasn't due to space / code complication trade-off vs time / elegance.

Reply:
Don't mind me, I'm just mentioning the actual @scheibelp ...